File: face/geomajas-face-dojo/client/src/main/resources/geomajas/gfx/ShapeStyle.js
dojo.provide("geomajas.gfx.ShapeStyle");
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
dojo.declare("ShapeStyle", null, {
	
	// defaults
	fillColor : "#FFFFFF",
	fillOpacity : "1",
	strokeColor : "#000000",
	strokeOpacity : "1",
	strokeWidth : "1",
	dashArray : null,
	symbol : null,	

	/**
	 * @fileoverview Simple style object for CSS.
	 * @class Simple style object for CSS.
	 * @author Pieter De Graef
	 *
	 * @constructor
	 */
	constructor : function (fillColor, fillOpacity, strokeColor, strokeOpacity, strokeWidth, dashArray, symbol) {
		if(fillColor != null){
			this.fillColor = fillColor;
		}
		if(fillOpacity != null){
			this.fillOpacity = fillOpacity;
		}
		if(strokeColor != null){
			this.strokeColor = strokeColor;
		}
		if(strokeOpacity != null){
			this.strokeOpacity = strokeOpacity;
		}
		if(strokeWidth != null){
			this.strokeWidth = strokeWidth;
		}
		if(dashArray != null){
			this.dashArray = dashArray;
		}
		if(symbol != null){
			this.symbol = symbol;
		}
	},

	set : function (fillColor, fillOpacity, strokeColor, strokeOpacity, strokeWidth, dashArray, symbol) {
		this.fillColor = fillColor;
		this.fillOpacity = fillOpacity;
		this.strokeColor = strokeColor;
		this.strokeOpacity = strokeOpacity;
		this.strokeWidth = strokeWidth;
		this.dashArray = dashArray;
		this.symbol = symbol;
	},

	clone : function () {
		return new ShapeStyle(this.fillColor, this.fillOpacity, this.strokeColor, this.strokeOpacity, this.strokeWidth, this.dashArray, this.symbol);
	},
	
	// copy non-null elements only !
	copyFrom : function (other) {
		if(other.fillColor != null){
			this.fillColor = other.fillColor;
		}
		if(other.fillOpacity != null){
			this.fillOpacity = other.fillOpacity;
		}
		if(other.strokeColor != null){
			this.strokeColor = other.strokeColor;
		}
		if(other.strokeOpacity != null){
			this.strokeOpacity = other.strokeOpacity;
		}
		if(other.strokeWidth != null){
			this.strokeWidth = other.strokeWidth;
		}
		if(other.dashArray != null){
			this.dashArray = other.dashArray;
		}
		if(other.symbol != null){
			this.symbol = other.symbol;
		}
	},
	
	getStrokeWidthAsFloat : function () {
		return parseFloat(this.strokeWidth);
	},

	// Getters and setters:

	getFillColor : function () {
		return this.fillColor;
	},

	setFillColor : function (fillColor) {
		this.fillColor = fillColor;
	},

	getFillOpacity : function () {
		return this.fillOpacity;
	},

	setFillOpacity : function (fillOpacity) {
		this.fillOpacity = fillOpacity;
	},

	getStrokeColor : function () {
		return this.strokeColor;
	},

	setStrokeColor : function (strokeColor) {
		this.strokeColor = strokeColor;
	},

	getStrokeOpacity : function () {
		return this.strokeOpacity;
	},

	setStrokeOpacity : function (strokeOpacity) {
		this.strokeOpacity = strokeOpacity;
	},

	getStrokeWidth : function () {
		return this.strokeWidth;
	},

	setStrokeWidth : function (strokeWidth) {
		this.strokeWidth = strokeWidth;
	},

	getDashArray : function () {
		return this.dashArray;
	},

	setDashArray : function (dashArray) {
		this.dashArray = dashArray;
	},

	getSymbol : function () {
		return this.symbol;
	},

	setSymbol : function (symbol) {
		this.symbol = symbol;
	}
});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/gfx/style/ShapeStyle.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.smartgwt.client.gfx.style;

import org.geomajas.configuration.FeatureStyleInfo;

import java.io.Serializable;

/**
 * <p>
 * Style object for shapes and geometries on the map.
 * </p>
 * 
 * @author Pieter De Graef
 */
public class ShapeStyle implements Style, Serializable {

	private static final long serialVersionUID = 150L;

	/**
	 * The color to be used as background color. Must be an HTML color code (red = #FF0000).
	 */
	private String fillColor;

	/**
	 * Value between 0 and 1, where 0 is completely transparent, and 1 means no transparency at all. This transparency
	 * is used for the background of the object on which it is applied.
	 */
	private float fillOpacity = -1;

	/**
	 * The color to be used as color for the edges. Must be an HTML color code (red = #FF0000).
	 */
	private String strokeColor;

	/**
	 * Value between 0 and 1, where 0 is completely transparent, and 1 means no transparency at all. This transparency
	 * is used for the edges/lines of the object on which it is applied.
	 */
	private float strokeOpacity = -1;

	/**
	 * The width in pixels of the edges of the object on which this style is to be applied.
	 */
	private float strokeWidth = -1;

	// Constructors:

	/**
	 * The default constructor. Leaves every field as is.
	 */
	public ShapeStyle() {
		// do nothing
	}

	/**
	 * Initialize this style object with all it's fields.
	 *
	 * @param fillColor fill color
	 * @param fillOpacity fill opacity
	 * @param strokeColor stroke color
	 * @param strokeOpacity stroke opacity
	 * @param strokeWidth stroke width
	 */
	public ShapeStyle(String fillColor, float fillOpacity, String strokeColor, float strokeOpacity, int strokeWidth) {
		this.fillColor = fillColor;
		this.fillOpacity = fillOpacity;
		this.strokeColor = strokeColor;
		this.strokeOpacity = strokeOpacity;
		this.strokeWidth = strokeWidth;
	}

	/**
	 * Initialize this style object with the values of another style object.
	 * 
	 * @param other
	 *            The other style object from which to copy the field values.
	 */
	public ShapeStyle(ShapeStyle other) {
		this.fillColor = other.fillColor;
		this.fillOpacity = other.fillOpacity;
		this.strokeColor = other.strokeColor;
		this.strokeOpacity = other.strokeOpacity;
		this.strokeWidth = other.strokeWidth;
	}

	/**
	 * Initialize this style object with the values of a style configuration object.
	 * 
	 * @param info
	 *            {@link FeatureStyleInfo} object to read the style from
	 */
	public ShapeStyle(FeatureStyleInfo info) {
		if (null != info) {
			this.fillColor = info.getFillColor();
			this.fillOpacity = info.getFillOpacity();
			this.strokeColor = info.getStrokeColor();
			this.strokeOpacity = info.getStrokeOpacity();
			this.strokeWidth = info.getStrokeWidth();
		}
	}

	// Other functions:

	/**
	 * Scales only the strokewidth.
	 */
	public void scale(double scale) {
		strokeWidth = (int) Math.ceil(strokeWidth * scale);
	}

	public void merge(ShapeStyle style) {
		if (style.getFillColor() != null) {
			fillColor = style.getFillColor();
		}
		setFillOpacity(style.getFillOpacity());
		if (style.getStrokeColor() != null) {
			strokeColor = style.getStrokeColor();
		}
		setStrokeOpacity(style.getStrokeOpacity());
		setStrokeWidth(style.getStrokeWidth());
	}

	/**
	 * Return a clone of this style object.
	 */
	public Style clone() { // NOSONAR super.clone() not supported by GWT
		return new ShapeStyle(this);
	}

	// Getters and setters:

	public String getFillColor() {
		return fillColor;
	}

	public void setFillColor(String fillColor) {
		this.fillColor = fillColor;
	}

	public float getFillOpacity() {
		return fillOpacity;
	}

	public void setFillOpacity(float fillOpacity) {
		if (fillOpacity >= 0 && fillOpacity <= 1) {
			this.fillOpacity = fillOpacity;
		}
	}

	public String getStrokeColor() {
		return strokeColor;
	}

	public void setStrokeColor(String strokeColor) {
		this.strokeColor = strokeColor;
	}

	public float getStrokeOpacity() {
		return strokeOpacity;
	}

	public void setStrokeOpacity(float strokeOpacity) {
		if (strokeOpacity >= 0 && strokeOpacity <= 1) {
			this.strokeOpacity = strokeOpacity;
		}
	}

	public float getStrokeWidth() {
		return strokeWidth;
	}

	public void setStrokeWidth(float strokeWidth) {
		if (strokeWidth >= 0) {
			this.strokeWidth = strokeWidth;
		}
	}
}